### PR TITLE
[REVIEW] Added general column comparison code for testing

### LINKED ISSUE #1387: [FEA] Some (trivial) utility functions for gdf_column's
There are some routine, mostly-trivial tasks we occasionally, or often, need to perform on `gdf_column`'s, the code for which is currently at best repeated and at worst partially-repeated in different places in libcudf. Also, when we use this code, we don't quite say what it is that we're doing:

* Ensuring a column is valid (e.g. not having more nulls than elements and a bunch of other checks)
* Determining whether two columns have matching types (which includes the nullability check) 
* Checking whether a column is nullable or not not. Right now we use the "implementation" of this check, i.e. check whether `some_column->valid == nullptr`, when we could say `is_nullable(some_column)`
* Determining whether two columns have compatible extra_type_info (which sometimes means doing nothing, depending on the dtype)

Let's have these utility functions implemented.

### PR DESCRIPTION
**Pull request by eyalroz**
This fixes issue #1367 ; it relies on PR #1390, which introduces some column utility functions (and fixes #1387). When this is merged it should allow merging PR #1428.

In addition to one file with the comparison functions, I've switched one of the tests - in `gather_tests.cu` over to use this facility, as an example/template of how to do so elsewhere.

To clarify:  This **can** be reviewed now, but **mustn't** be merged until PR #1390 is merged (at which point I will rebase).

### CODE REVIEW

**Comment by harrism** [comment]
> To clarify: This **can** be reviewed now, but **mustn't** be merged until PR #1390 is merged (at which point I will rebase).

Please `git merge branch-0.8`. Don't rebase.

**Comment by harrism** [comment]
What is wrong with `column_wrapper::operator==`? It is nice to add utilities to print only the mismatching elements, but couldn't you build on the existing comparison operator?

**Comment by eyalroz** [comment]
> Please `git merge branch-0.8`. Don't rebase.

Ok.

> What is wrong with `column_wrapper::operator==`? It is nice to add utilities to print only the mismatching elements, but couldn't you build on the existing comparison operator?

Adapting and expanding my [answer to that question on the issue page](https://github.com/rapidsai/cudf/issues/1367#issuecomment-481391251):

1. Inconsistent with the behavior of things which aren't columns (including, say, `gdf_scalar`) on comparison
2. Inconsistent with the behavior of `gdf_column` (for which, in some cases, comparison is also relevant). While it's true `gdf_column` will be redesigned/supreseded - the column wrapper class is also likely to undergo some [changes](https://github.com/rapidsai/cudf/issues/1178).
3. Breaks the element of least surprise. It's surprising for `col_a == col_b` to print things.
4. Inconsistent with the pattern of using gtest; we are used to saying `EXPECT_...` or `ASSERT_` rather than have the effect of those just sort of happen on the side/behind the scenes. <- TBH, my function is a bit inconsistent with that as well, but I at least tried to minimize this.
5. Classes should, in general, have the minimum functionality they need to; and `column_wrapper` is already packed with functionality I'm not sure it even should have.
6. `operator==` cannot be (easily) parameterized, while my code already has one parameter, and will likely develop more (e.g. floating point inequality threshold, print relative or absolute differences, compare bitwise, maybe some parameters regarding printing, etc.)

File: cpp/tests/utilities/compare_column_wrappers.cuh [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    const gdf_column& rhs_gdf_column = *(rhs.get());
+    cudf::validate(lhs_gdf_column);
+    cudf::validate(rhs_gdf_column);
+    EXPECT_TRUE(cudf::have_same_type(lhs_gdf_column, rhs_gdf_column));
+    if (not cudf::have_same_type(lhs_gdf_column, rhs_gdf_column)) { return; }
+    EXPECT_EQ(lhs_gdf_column.size, rhs_gdf_column.size);
+    EXPECT_EQ(lhs_gdf_column.null_count, rhs_gdf_column.null_count);
+    auto common_size = lhs_gdf_column.size;
+    if (common_size == 0) { return; }
+    auto non_nullable = (lhs_gdf_column.null_count == 0);
```
Comment (jrhemstad) [inline cpp/tests/utilities/compare_column_wrappers.cuh]: Should this use `is_nullable` or `has_nulls` instead?

File: cpp/tests/utilities/compare_column_wrappers.cuh [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+{
+    auto lhs_non_nullable = (lhs_validity_on_host == nullptr);
+    auto rhs_non_nullable = (rhs_validity_on_host == nullptr);
+    auto max_name_length = std::max(lhs_name.length(), rhs_name.length());
+
+    for(gdf_size_type i = 0; i < common_size; i++) {
+        auto lhs_element_is_valid = lhs_non_nullable or gdf::util::bit_is_set<gdf_valid_type, gdf_size_type>(lhs_validity_on_host, i);
+        auto rhs_element_is_valid = rhs_non_nullable or gdf::util::bit_is_set<gdf_valid_type, gdf_size_type>(rhs_validity_on_host, i);
+        auto elements_are_equal =
+            (not lhs_element_is_valid and not rhs_element_is_valid) or
```
Comment (jrhemstad) [inline cpp/tests/utilities/compare_column_wrappers.cuh]: Document that this treats `null == null`. Perhaps even parameterize this behavior.

File: cpp/tests/utilities/compare_column_wrappers.cuh (hunk 1 above)
Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: Actually, on second though, this should exist at all. See upcoming commit.

File: cpp/tests/utilities/compare_column_wrappers.cuh (hunk 2 above)
Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: Good catch. Parameterized (in upcoming commit).

File: cpp/tests/utilities/compare_column_wrappers.cuh [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <iomanip>
+#include <tuple>
+
+namespace detail {
+
+template <typename T>
```
Comment (jrhemstad) [inline cpp/tests/utilities/compare_column_wrappers.cuh]: How about instead:

```
template <typename T>
std::string to_string(T val) {
  return std::to_string(cudf::detail::unwrap(val));
}

std::string to_string(cudf::bool8 val) {
  return {cudf::detail::unwrap(val) ? "true" : "false"};
}
```

Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: Ok, sure.

Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: Ah, right, `unwrap()` is a pass-through for non-wrapped types. Sure.

File: cpp/tests/utilities/compare_column_wrappers.cuh [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    cudf::test::column_wrapper<E> const&  rhs,
+    const std::string&                    rhs_name,
+    bool                                  treat_distinct_nulls_as_equal = distinct_nulls_are_equal,
+    bool                                  print_all_unequal_pairs = dont_print_unequal_pairs)
+{
+    const gdf_column& lhs_gdf_column = *(lhs.get());
+    const gdf_column& rhs_gdf_column = *(rhs.get());
+    cudf::validate(lhs_gdf_column);
+    cudf::validate(rhs_gdf_column);
+    EXPECT_TRUE(cudf::have_same_type(lhs_gdf_column, rhs_gdf_column));
```
Comment (jrhemstad) [inline cpp/tests/utilities/compare_column_wrappers.cuh]: I think it'd be nice to know _why_ two columns don't have the same type, and so I'd move the body of `have_same_type` to here with proper `EXPECT_EQ`, e.g.:
```
    EXPECT_EQ(lhs_gdf_column.dtype, rhs_gdf_column.dtype);
    EXPECT_EQ(is_nullable(lhs_gdf_column), is_nullable(rhs_gdf_column));
    if (lhs_gdf_column.dtype == GDF_TIMESTAMP) {
      EXPECT_EQ(lhs_gdf_column.dtype_info.time_unit,
                lhs_gdf_column.dtype_info.time_unit);
    }
```

Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: Maybe I'll just write an `expect_same_type()` function then.

Comment (eyalroz) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: How about:
```
// Note: This has quite a bit of code repretition, plus it repeats all of `cudf::util::have_same_type` essentially.
bool expect_columns_have_same_type(const gdf_column& validated_column_1, const gdf_column& validated_column_2, bool ignore_extra_type_info = false) noexcept
{
    EXPECT_EQ(validated_column_1.dtype, validated_column_2.dtype);
    if(validated_column_1.dtype != validated_column_2.dtype) { return false; }
    EXPECT_EQ(cudf::is_nullable(validated_column_1), cudf::is_nullable(validated_column_2));
    if (cudf::is_nullable(validated_column_1) != cudf::is_nullable(validated_column_2)) { return false; }
    if (ignore_extra_type_info) { return true; }
    auto common_dtype = validated_column_1.dtype;
    EXPECT_TRUE(cudf::detail::extra_type_info_is_compatible(common_dtype, validated_column_1.dtype_info, validated_column_2.dtype_info));
    return cudf::detail::extra_type_info_is_compatible(common_dtype, validated_column_1.dtype_info, validated_column_2.dtype_info);
}
```

Comment (jrhemstad) [reply cpp/tests/utilities/compare_column_wrappers.cuh]: That looks great.

File: cpp/tests/utilities/compare_column_wrappers.cuh [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    bool                                  print_all_unequal_pairs = dont_print_unequal_pairs)
+{
+    const gdf_column& lhs_gdf_column = *(lhs.get());
+    const gdf_column& rhs_gdf_column = *(rhs.get());
+    cudf::validate(lhs_gdf_column);
+    cudf::validate(rhs_gdf_column);
+    expect_columns_have_same_type(lhs_gdf_column, rhs_gdf_column);
+    if (not cudf::have_same_type(lhs_gdf_column, rhs_gdf_column)) { return; }
+    EXPECT_EQ(lhs_gdf_column.size, rhs_gdf_column.size);
+    EXPECT_EQ(cudf::is_nullable(lhs), cudf::is_nullable(rhs));
```
Comment (jrhemstad) [inline cpp/tests/utilities/compare_column_wrappers.cuh]: This is already checked in `expect_columns_have_same_type`.

**Comment by eyalroz** [comment]
@jrhemstad : Merge time...

**Pull request merged**